langchain/vectorstores/matching_engine.py

```diff
@@ -0,0 +1,465 @@
+
+        from google.cloud import storage
+
+        if self.gcs_client is None:            
+            self.gcs_client = storage.Client(
+                credentials=self.credentials, 
+                project=self.project_id
+            )
+
+        return self.gcs_client
+
+    def similarity_search(
+        self, 
+        query: str, 
+        k: int = 4, 
+        **kwargs: Any
+    ) -> List[Document]:
+        """Return docs most similar to query.
+
+        Args:
+            query: The string that will be used to search for similar documents.
+            k: The amount of neighbors that will be retrieved.
+
+        Returns:
+            A list of k matching documents.
+        """
+
+        logger.debug(f"Embedding query {query}.")
+        embedding_query = self.embedding.embed_documents([query])
+
+        index_id = None
+
+        for index in self.endpoint.deployed_indexes:
+            if index.index == self.index.resource_name:
+                index_id = index.id
+                break
+        
+        if index_id == None:
+            raise ValueError(f"No index with id {self.index.resource_name} "
+                             f"deployed on enpoint "
+                             f"{self.endpoint.display_name}.")
+
+        response = self.endpoint.match(
+            deployed_index_id=index_id,
+            queries=embedding_query,
+            num_neighbors=k,
+        )
+
+        if len(response) == 0:
+            return []
+
+        logger.debug(f"Found {len(response)} matches for the query {query}.")
+
+        results = []
+
+        # I'm only getting the first one because queries receives an array
+        # and the similarity_search method only recevies one query. This
+        # means that the match method will always return an array with only
+        # one element.
+        for doc in response[0]:
+            print(doc)
+            page_content = self._download_from_gcs(f"documents/{doc.id}")
+            results.append(Document(page_content=page_content))
+
+        logger.debug(f"Downloaded documents for query.")
+            
+        return results
+    
+    def _download_from_gcs(self, gcs_location: str) -> str:
+        """Downloads from GCS in text format.
+
+        Args:
+            gcs_location: The location where the file is located.
+
+        Returns:
+            The string contents of the file.
+        """
+        client = self._get_gcs_client()
+        bucket = client.get_bucket(self.gcs_bucket_uri)
+        blob = bucket.blob(gcs_location)
+        return blob.download_as_string()
+    
+    @classmethod
+    def from_texts(
+        cls: Type["MatchingEngine"], 
+        texts: List[str],
+        embedding: Embeddings = None,
+        metadatas: Optional[List[dict]] = None, 
+        project_id: str = None,
+        region: str = None,
+        gcs_bucket_uri: str = None,
+        index_id: str = None,
+        endpoint_id: str = None,
+        credentials_path: str = None,
```

Comment:
If they're typed as `str` we shouldn't have a default of `None`. If we want these to be required keyword arguments, we could  do the patterh

```
def foo(*, region: str, index_id: str):
```

(the lone asterisk)